File: python/bracket-push/bracket_push.py
def is_paired(input_string):
  bracket_stack = []
  for char in input_string:
    # if opening bracket, add match to stack
    if char == '[':
      bracket_stack.append(']')
    elif char == '{':
      bracket_stack.append('}')
    elif char == '(':
      bracket_stack.append(')')
    elif char in ']})':
      # closing bracket without any brackets or unmatched
      if len(bracket_stack) == 0 or bracket_stack.pop() != char:
        return False
  # check if any brackets are still on stack
  return len(bracket_stack) == 0

File: cpp/bracket-push/bracket_push.cpp
#include "bracket_push.h"
#include <vector>

bool bracket_push::check(std::string brackets) {
  std::vector<char> stack;
  for (size_t i = 0; i < brackets.length(); i++) {
    char c = brackets[i];
    if (c == '[') {
      stack.push_back(']');
    } else if (c == '{') {
      stack.push_back('}');
    } else if (c == '(') {
      stack.push_back(')');
    } else if (c == ']' || c == '}' || c == ')') {
      if (stack.size() == 0 || stack[stack.size()-1] != c) {
        return false;
      }
      stack.pop_back();
    }
  }
  return stack.size() == 0;
}